### DISCUSSION TITLE
[11/08] 이경준 / Hiddenite

### DISCUSSION THREAD

**Discussion by k-jun-lee** [sjquan/2022-Study#25] [Seminars]
11/08 발표에 대한 Q&A입니다.

**Comment by k-jun-lee** [comment]
Q: 최근 ViT 계열에서 pruning 연구가 많이 진행되고 있는 것 같습니다. 근데 Transformer 계열의 모델을 프루닝할때 인풋마다 dynamic 한 구조가 만들어져서 실제 가속하는데 오버헤드가 있을 거 같습니다. 혹시 이러한 오버헤드를 해결하기 위한 하드웨어 관련 연구가 이뤄지고 있을까요?

**Reply by k-jun-lee** [reply]
A: ViT 계열의 Pruning에 대해서 저는 잘 알지 못해서 이와 관련하여 어떤 hardware requirement가 있는지 잘 알지 못합니다. Dynamic한 구조라는 것이 정확하게 어떤 구조를 지칭하시는지 잘 모르겠어서 현재 질문으로는 답변드리기가 어려울 것 같은데요. 관련된 paper를 알려주시면 살펴보고 답변 드릴 수 있는 선에서 답변드릴 수 있도록 하겠습니다.

**Comment by k-jun-lee** [comment]
Q: 발표 중에도 질문한 내용이긴한데, lottery ticket이나 What's hidden in a randomly weighted NNs에서와 같이 큰 모델안에 좋은 initialized weights가 있다는 논문중에서는 학습된 batch norm parameters(beta, gamma, mean, variance)를 사용하지않고 초기값으로 사용하는지 궁금합니다.

**Reply by k-jun-lee** [reply]
A: What's hidden in a randomly weighted neural network? 논문의 내용으로 답변을 대신하겠습니다. back propagation 통해 학습되는 
parameter인 beta/gamma는 학습하지 않습니다, mean/variance에 대해서는 학습하지 않는다는 명시적인 언급이 없고, 통계값이기 때문에 계산하는 것으로 보입니다.

" For simplicity, our edge-popup algorithm does not modify batch norm parameters, they are frozen at their default initialization in PyTorch (i.e. bias 0,scale 1)"

**Comment by k-jun-lee** [comment]
Q: 제가 하드웨어나 컴파일러에 대해 무지해서 하는 질문 일 수도 있는데, (발표 끝나고 여러가지로 오갔던 이야기들과 비슷한 질문 인데 제가 답변을 잘 이해를 못해서요;) unstructued pruning은 GPU에서는 병렬성이 깨어져서 FLOP만 좋아지지 실제로는 throughput에서 향상이 없다 라고 알고 있는데, 지금은 Dram access하는게 적어져서 속도 향상이 있다. 라고 보면 되는 건가요?"

**Reply by k-jun-lee** [reply]
A: 말씀해주신 대로 irregular pattern을 가지는 pruning은 throughput을 저하시키는 것이 맞습니다. throughput에 관련된 내용은 논문(Hiddenite)에 나와있지 않아서 저도 궁금하네요. 논문은 1) DRAM Access를 줄였고, 2)"연산당 에너지 효율(TOPS/W)"이라는 metric를 크게 향상시킨 논문으로 봐주시면 될 것 같습니다.

**Comment by k-jun-lee** [comment]
Q: 어차피 랜덤한 weight들을 이용해서 학습할거라면 모든 층에 대해서 weight를 random하게 새로 생성하는 것보다 하나의 weight에서 mask를 생성하는 방법도 생각해볼 수 있을 것 같습니다. 하지만, 이 방법의 경우 tiling (?) 방법에 따라 연산할 때 어차피 한 타일 계산하고 다음 층에서 해당 값을 사용할 때 그 타일을 새로 생성해서 계산해야해서 연산적인 이득이 없을 것 같은데 혹시 맞을까요? 다르게 말하자면 모든 층에 대해서 weight를 공유하고 mask만을 다르게 했을 때 하드웨어적인 이득이 있을 지 궁금합니다. (성능은 우선 제외하고요)

**Reply by k-jun-lee** [reply]
A: weight를 random하게 생성해도 좋은 네트워크를 만들 수 있기 때문에 weight를 생성하는 것을 굉장히 간단 하드웨어 유닛인 RNG(random number generator)로 대체할 수 있었습니다.  말씀해주신 "random하게 새로 생성하는 것보다 하나의 weight에서 mask를 생성하는 방법"은 논문에서 말하는 내용과 동일한 것으로 생각됩니다. 한편, 마지막 질문 주신 내용에서 마지막 문장 "다르게 말하자면 모든 층에 대해서 weight를 공유하고 mask만을 다르게 했을 때 하드웨어적인 이득이 있을 지 궁금합니다."에서 하드웨어적인 이득이라는 것은 어떤 부분을 말씀하지는 것인지도 애매해서 답변을 드리기가 모호하네요. 제가 질문을 잘 이해하지 못한 것 같은 느낌이 있어서 구체적으로 질문해주시면 다시 답변 드리겠습니다!

**Comment by k-jun-lee** [comment]
Q: subnet을 찾기 위한 원래 network자체가 resnet처럼 overparameterized 되어서 redundant한 weight들이 많은 모델들 말고 MobileNet 등과 같이 5M 이하의 parameter들을 가지는 것들에 대해서도 lottery ticket을 찾을 수 있는지 궁금합니다.

**Reply by k-jun-lee** [reply]
A: 해당 부분에 대해서는 저도 직접 구현/실험을 해보지 않아서 딱 잘라서 답변드리기가 어렵습니다. 다만, 소개드린 논문 "What's hidden in a randomly weighted neural network?"에서 2, 4, 6, 8 layer CNN에 대한 실험이 있기 때문에 적은 파라미터를 가지는 네트워크에서도 가능할 것라고 예상해볼 수 있을 것 같습니다.

**Comment by k-jun-lee** [comment]
Q: 슈퍼 마스크를 통해 나오는 필터는 값이 랜덤한 위치에 해당되는 Unstructured 형태인 것으로 생각됩니다. 보통은 이러한 경우에 Gating 회로를 추가하여 해결하고 있습니다. 논문에서는 Data gating이라는 방법을 사용한다고 하는데 이에 대한 설명이 있으면 좋을 것 같습니다.

**Reply by k-jun-lee** [reply]
A: 특별한 것은 아니고요. 말씀해주신 일반적인 Gatining 회로랑 동일한테 Mask에 따라 연산기(PE)를 동작시키지 않는 구조입니다. 하드웨어 레벨에서는 logic value가 0->1 혹은 1->0으로 변할 때 current가 발생해서 power가 소모됩니다. 논문에서는 feature는 8-bit을 사용하고 weight는 1-bit을 사용하고 있기 때문에 8-bit feature에 대한 변화를 줄이기 위해서 feature에 대한 변화를 줄이기 위해 data(=feature) gating이라고 표현해둔 것 같습니다.

**Comment by k-jun-lee** [comment]
Q: 개인적으로 하드웨어에서 모델에 따라 설계하는 방법에 대해 공부하고 싶은데, 독학이 쉽지가 않네요. 독학하기에 좋은 방법을 알려주시면 큰 도움이 될 것 같습니다.

**Reply by k-jun-lee** [reply]
A: 제가 이런 조언을 드릴 수 있는 사람인가라는 생각이 들어서 답변을 드리기 조심스럽지만, 혼자서 project를 하나 해보는 것이 가장 좋은 방법이라고 생각이 됩니다. 다만 현실적으로 불가능한 경우가 많을 것이기 때문에 간단한 자료라도 원하신다면 아래사이트를 시작포인트로 잡는 것도 괜찮을 것 같습니다.

http://eyeriss.mit.edu/tutorial.html

**Comment by k-jun-lee** [comment]
Q: Supermask pruning이 structured pruning 인지 unstructured pruning인지 아직도 헷갈립니다. Structured pruning이 아니라면 sparse matrix 표현으로 바꾸지 않는다면 오히려 HW 가속기 제작에 어려움이 있기 때문입니다. Hidden network 논문 figure 2를 보면 일부 weight 만 없애는 unstructured pruning으로 보이는데, 발표에선 또 아니라고 하시니 어떤게 정확히 맞는 프루닝 방식인지 궁금합니다.

**Reply by k-jun-lee** [reply]
A: 각 weight 마다 score를 가지고 있고, 이 score에 따라 pruning을 하기 때문에 말씀해주신 것처럼 unstructured pruning이 맞습니다. 발표 영상을 다시 보는데 제가 긴장을 해서인지 횡설수설이 있었네요 :-<

**Comment by k-jun-lee** [comment]
Q: Cuda와 같은 low-level 프로그래밍 하는 사람들에게 추천해주실만한 페이퍼가 있는지 궁금합니다. 하드웨어적인 내용도 좋습니다. 하드웨어쪽 페이퍼들을 잘 모르다보니 큰 도움이 될 것 같습니다.

**Reply by k-jun-lee** [reply]
A: 제가 잘 알지 못하는 분야이기도 하고, 질문의 내용도 조금 포괄적이라 답변을 드리기가 어렵네요 :) 조금 더 구체적으로 질문해주시면 아는 선에서 답변 드리겠습니다!
